p4include/psa.p4

```diff
@@ -0,0 +1,675 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _PORTABLE_SWITCH_ARCHITECTURE_P4_
+#define _PORTABLE_SWITCH_ARCHITECTURE_P4_
+
+/**
+ *   P4-16 declaration of the Portable Switch Architecture
+ */
+
+/**
+ * These types need to be defined before including the architecture file
+ * and the macro protecting them should be defined.
+ */
+#ifndef PSA_CORE_TYPES
+#error "Please define the following types for PSA and the PSA_CORE_TYPES macro"
+typedef bit<unspecified> PortId_t;
+const   PortId_t         PORT_CPU = unspecified;
+
+typedef bit<unspecified> MulticastGroup_t;
+typedef bit<unspecified> PacketLength_t;
+typedef bit<unspecified> EgressInstance_t;
+typedef bit<unspecified> ParserStatus_t;
+typedef bit<unspecified> ParserErrorLocation_t;
+typedef bit<unspecified> entry_key;           /// for DirectCounters
+
+// typedef bit<unspecified> InstanceType_t;
+// const   InstanceType_t   INSTANCE_NORMAL = unspecified;
+#endif
+
+#include "core.p4"
+
+/* --------------------------------------------------------------------- *
+/**
+ * PSA supported metadata types
+ */
+typedef enum { NORMALINSTANCE_, CLONE_INSTANCE } InstanceType_t;
+
+struct psa_parser_input_metadata_t {
+  PortId_t                 ingress_port;
+  InstanceType_t           instance_type;
+}
+
+struct psa_ingress_input_metadata_t {
+  PortId_t                 ingress_port;
+  InstanceType_t           instance_type;  /// Clone or Normal
+  /// set by the runtime in the parser, these are not under programmer control
+  ParserStatus_t           parser_status;
+  ParserErrorLocation_t    parser_error_location;
+}
+
+struct psa_ingress_output_metadata_t {
+  PortId_t                 egress_port;
```

Comment:
Is it intended to have both an egress_port field here, _and_ the send_to_port() method in the PRE object?  It seems simpler to have only one of the two, probably the PRE method, and eliminate this egress_port field.  I could imagine having both, and specifying what happens if you both assign to the egress_port field and call send_to_port(), but why bother creating the confusion?